### PR TITLE
Blocks: Refactor the check against self for block hooks

### PR DESCRIPTION
**Pull request by gziolo**
Trac ticket: https://core.trac.wordpress.org/ticket/59346

I added some refactorings discussed in https://github.com/WordPress/wordpress-develop/pull/5203#pullrequestreview-1626828337:
- ~https://github.com/WordPress/wordpress-develop/pull/5203#discussion_r1326102815 - `gutenberg` text domain remove~ committed with https://core.trac.wordpress.org/changeset/56592.
- There is custom logic for mapping `blockHooks` to `block_hooks` when handling metadata. I covered that part in unit tests.
- One last bit of feedback was for the _doing_it_wrong logic. I moved the check to the `WP_Block_Type` class to cover also the case when someone registers the block with a regular `register_block_type` call without using metadata file.
---
**This Pull Request is for code review only. Please keep all other discussion in the Trac ticket. Do not merge this Pull Request. See [GitHub Pull Requests for Code Review](https://make.wordpress.org/core/handbook/contribute/git/github-pull-requests-for-code-review/) in the Core Handbook for more details.**

### CODE REVIEW

File: src/wp-includes/class-wp-block-type.php [hunk 1]
```diff
@@ -522,6 +522,20 @@ public function set_props( $args ) {
...
+				}
+				if ( array_key_exists( $this->name, $property_value ) ) {
+					_doing_it_wrong(
+						__METHOD__,
+						__( 'Cannot hook block to itself.' ),
+						'6.4.0'
+					);
+					unset( $property_value[ $this->name ] );
+				}
+			}
```
Comment (ockham) [inline src/wp-includes/class-wp-block-type.php]: Sorry, I really think we should keep this in `register_block_type_from_metadata` 😬 AFAICS, no other field is being validated here in `WP_Block_Type` -- they're all processed in `register_block_type_from_metadata`, so I'd rather keep it consistent.

Note that `register_block_type` actually calls `register_block_type_from_metadata` for block registration, so any block type registered via `register_block_type` will in fact undergo the validation in `register_block_type_from_metadata`: https://github.com/WordPress/wordpress-develop/blob/e005108fe18e6278f4e6dd0fec030e503fe0ff46/src/wp-includes/blocks.php#L584-L590

Finally, note that there's even an "inner" inconsistency with this PR, as validation of the allowed relative positions is still done in `register_block_type_from_metadata` 😅

File: tests/phpunit/tests/blocks/register.php [hunk 2]
```diff
@@ -645,8 +645,13 @@ public function test_block_registers_with_metadata_fixture() {
...
 		// @ticket 59346
-		$this->assertSame(
-			array( 'core/post-content' => 'before' ),
+		$this->assertSameSets(
+			array(
+				'tests/before'      => 'before',
+				'tests/after'       => 'after',
+				'tests/first-child' => 'first_child',
+				'tests/last-child'  => 'last_child',
+			),
```
Comment (ockham) [inline tests/phpunit/tests/blocks/register.php:654]: 👍

**Review by ockham — CHANGES_REQUESTED** [review]
See my comment -- I really think we should keep the check for self-insertion in `register_block_type_from_metadata`.

**Comment by gziolo** [comment]
> See my comment -- I really think we should keep the check for self-insertion in `register_block_type_from_metadata`.

I tried locally reverting the changes for the code moved to `WP_Block_Type::set_props`. I basically used the current `trunk` with the same unit test but modified `@expectedIncorrectUsage` line:

```diff
diff --git a/tests/phpunit/tests/blocks/register.php b/tests/phpunit/tests/blocks/register.php
index 55faf297a5..2c1cf64937 100644
--- a/tests/phpunit/tests/blocks/register.php
+++ b/tests/phpunit/tests/blocks/register.php
@@ -1069,4 +1069,29 @@ class Tests_Blocks_Register extends WP_UnitTestCase {
 		$actual = register_block_style( 'core/query', $block_styles );
 		$this->assertTrue( $actual );
 	}
+
+	/**
+	 * @ticket 59346
+	 *
+	 * @covers ::register_block_type
+	 *
+	 * @expectedIncorrectUsage register_block_type_from_metadata
+	 */
+	public function test_register_block_hooks_targeting_itself() {
+		$block_name = 'tests/block-name';
+		$block_type = register_block_type(
+			$block_name,
+			array(
+				'block_hooks' => array(
+					$block_name         => 'first',
+					'tests/other-block' => 'last',
+				),
+			)
+		);
+
+		$this->assertSameSets(
+			array( 'tests/other-block' => 'last' ),
+			$block_type->block_hooks
+		);
+	}
 }
```

The test fails:

Screenshot 2023-09-15 at 09 01 14

No changes get applied.

---

When inspecting the code it turns out that `register_block_type_from_metadata` isn't the universal way to register block type:

https://github.com/WordPress/wordpress-develop/blob/227fdb8b741a71e300062d80e9471fef87a5ff11/src/wp-includes/blocks.php#L618-L624

If no path to the `block.json` file gets passed as the first param, then the registration goes directly through the `WP_Block_Type_Registry`. It's also what `register_block_type_from_metadata` calls after all transformations applied to the loaded `block.json` file:

https://github.com/WordPress/wordpress-develop/blob/227fdb8b741a71e300062d80e9471fef87a5ff11/src/wp-includes/blocks.php#L594-L597

We can keep the implementation as it is currently works in `trunk` and refactor the test to use some mocked `block.json` file. The drawback is that `_doing_it_wrong` check won't run in the case when someone opts into registering the block type using the syntax that has been present in WordPress core since version 5.0. The same applies to the case when someone uses the registry directly with `WP_Block_Type_Registry::get_instance()->register()`. Let me know which approach you prefer, given all the above.

File: src/wp-includes/class-wp-block-type.php (hunk 1 above)
Comment (gziolo) [reply src/wp-includes/class-wp-block-type.php]: I left a https://github.com/WordPress/wordpress-develop/pull/5218#issuecomment-1720779818 with the details from my debugging after trying to use the test with the version before refactoring.

> Finally, note that there's even an "inner" inconsistency with this PR, as validation of the allowed relative positions is still done in register_block_type_from_metadata 😅

Yes, that could get improved, but we first need to settle on the responsibilities of processing `block.json` files. So far, it is mostly about mapping the JSON object to the matching fields in `WP_Block_Type`.

> AFAICS, no other field is being validated here in WP_Block_Type

There are some issues filed in Gutenberg to address it:
- https://github.com/WordPress/gutenberg/issues/48039
- https://github.com/WordPress/gutenberg/issues/48112

**Comment by ockham** [comment]
> When inspecting the code it turns out that `register_block_type_from_metadata` isn't the universal way to register block type:
> 
> https://github.com/WordPress/wordpress-develop/blob/227fdb8b741a71e300062d80e9471fef87a5ff11/src/wp-includes/blocks.php#L618-L624
> 
> If no path to the `block.json` file gets passed as the first param, then the registration goes directly through the `WP_Block_Type_Registry`. 

You're of course right 🤦‍♂️ My apologies, I'd read that code but totally glossed over the `if ( is_string( $block_type ) && file_exists( $block_type ) )` conditional.

> It's also what `register_block_type_from_metadata` calls after all transformations applied to the loaded `block.json` file:
> 
> https://github.com/WordPress/wordpress-develop/blob/227fdb8b741a71e300062d80e9471fef87a5ff11/src/wp-includes/blocks.php#L594-L597
> 
> We can keep the implementation as it is currently works in `trunk` and refactor the test to use some mocked `block.json` file. The drawback is that `_doing_it_wrong` check won't run in the case when someone opts into registering the block type using the syntax that has been present in WordPress core since version 5.0. The same applies to the case when someone uses the registry directly with `WP_Block_Type_Registry::get_instance()->register()`. Let me know which approach you prefer, given all the above.

I think since there's no precedent of doing any processing or validation in `WP_Block_Type`, I'd like to keep it that way, even if it means that people could bypass the check in `register_block_type_from_metadata`. In any case, I'd like to have a check that prevents infinite recursion by disallowing self-insertion later in `insert_hooked_block` (or whatever we're going to call it), regardless of how the block was registered.

**Comment by gziolo** [comment]
Thanks for the feedback @ockham, I will refactor the code accordingly.

**Review by ockham — APPROVED** [review]
Thank you! LGTM :shipit:

**Comment by gziolo** [comment]
Committed with https://core.trac.wordpress.org/changeset/56607.

**Comment by ockham** [comment]
Just a short note: The method we landed on for [inserting hooked blocks](https://github.com/WordPress/wordpress-develop/pull/5261) into their designated positions should make infinite loops of blocks inserting themselves next to each other impossible, if I'm not mistaken.

The crux is that we no longer insert parsed hooked blocks into the block tree, where they might be traversed themselves by `traverse_and_serialized`, and blocks that are hooked to _them_ would be inserted, potentially causing infinite recursion. Instead, we now only concatenate the hooked block's serialized markup during the traversal and serialization step to its anchor block's. That serialized markup is _not_ subjected to any further parsing or traversal that could cause recursive block insertion.

In other words, while it's nice to prohibit "self-insertion" at block type registration stage, it's not technically required -- things won't break if a block does end up registering itself for "self-insertion".

**Comment by gziolo** [comment]
> In other words, while it's nice to prohibit "self-insertion" at block type registration stage, it's not technically required -- things won't break if a block does end up registering itself for "self-insertion".

That's another excellent side-effect of the revised implementation. I really like how you integrated Block Hooks into WordPress core, taking advantage of the access to all internals. It seems fine to leave the validation as is. By the way, there is an open issue to add validation for every argument to `WP_Block_Type`: https://github.com/WordPress/gutenberg/issues/48039.
